### ISSUE TITLE
[FEATURE] Support NEFTune noisy embedding vectors technique

### ISSUE THREAD

**Issue by tmostak** [h2oai/h2o-llmstudio#492]
### 🚀 Feature

In this [paper](https://arxiv.org/pdf/2310.05914.pdf), the authors introduced a concept they called NEFTune, which basically shows that LLM downstream performance can be improved if noise is added to the embedding vectors during instruct fine-tuning. Could this capability be supported in LLM Studio? The capability has already been added to Huggingface's TRL (Transformer Reinforcement Library).

### Motivation

The promising results of the paper suggest it could be useful to have the ability to have this capability in LLM Studio to maximize downstream performance of LLMs trained in the platform.

Here is a link to their [repo](https://github.com/neelsjain/NEFTune).

**Comment by maxjeblick** [comment]
Thanks for posting this interesting paper! I have added nefttune in [this branch](https://github.com/h2oai/h2o-llmstudio/tree/max/add_nefttune), based upon trl's implementation.

It's concise enough that it should work, but it may still take a while for me fully test it (see TODOs).
Feel free to test the branch linked above, also very happy about feedback and potential model improvements!

**Comment by maxjeblick** [comment]
As a side note, I haven't seen this (or similar) methods to work in the past (==deberta finetuning); so I'm a bit cautious about the claims of the paper.

**Comment by tmostak** [comment]
Wow @maxjeblick, thanks for the amazingly quick turnaround, was not expecting that! 

I'm trying out the branch now and should have some comparison points to post soon for the text-to-sql model we're training (not sure this is the right use case for this technique, but at least it's a data point).

**Comment by tmostak** [comment]
Sorry for the delayed follow-up.

I've tested this in a variety of contexts using 7B Code Llama models with full fine-tunes, and found that in some cases I got slightly lower final eval perplexity using an NEFT alpha of 5, as well as 1-2 percentage points better accuracy on a downstream SQL generation task, although a few other models had the same or slightly worse performance with NEFT on, so  I'd need more rigorous testing to definitively tell if I'm seeing real benefit or just measuring noise at this point.

Still, it seems like a useful knob to have and seems to be robust (haven't hit any errors or other issues training 10+ models with it so far). I also should note that perhaps it has greater value for general-purpose text generation, which was what was explored in the original paper, versus highly constrained tasks like SQL generation, so my results should only be taken as one data point.

**Comment by maxjeblick** [comment]
Thanks for the detailed summary (good to hear the metrics improved) and bringing up NEFT in the first place!
I'll finish working on the branch and create a PR shortly.

**Comment by maxjeblick** [comment]
Hi @tmostak, sorry for the period of inactivity; I was occupied with other tasks the last weeks. I'll create a PR for this feature the next days.

**Comment by tmostak** [comment]
Thanks @maxjeblick for landing this!
